### PR TITLE
Fb1489

### PR DESCRIPTION
**Pull request by spireken**
There's still the issue with moving the mouse into the chart view and back into the color picker, but that will be fixed when the cursor case is fixed.

### CODE REVIEW

File: Applications/Spire/Include/Spire/Charting/ColorPicker.hpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+      //! Constructs a ColorPicker.
+      /*
+        \param width The width of the color picker image.
+        \param height The height of the color picker image.
+        \param parent The parent widget.
+      */
+      ColorPicker(int width, int height, QWidget* parent = nullptr);
+
+      boost::signals2::connection connect_preview_signal(
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/ColorPicker.hpp]: Document all public API methods.

File: Applications/Spire/Include/Spire/Charting/DropDownColorPicker.hpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#ifndef SPIRE_DROPDOWN_COLOR_PICKER_HPP
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/DropDownColorPicker.hpp]: Drop down is treated as two words.

File: Applications/Spire/Include/Spire/Charting/DropDownColorPicker.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    public:
+
+      //! Signals that a color has been selected.
+      using ColorSignal = Signal<void (const QColor& color)>;
+
+      //! Constructs a DropdownColorPicker.
+      /*
+        \param parent The parent widget.
+      */
+      DropDownColorPicker(QWidget* parent = nullptr);
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/DropDownColorPicker.hpp]: Mark single parameter constructors as ```explicit```.

File: Applications/Spire/Include/Spire/Charting/DropDownColorPicker.hpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+      boost::signals2::connection connect_color_signal(
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/DropDownColorPicker.hpp]: Document all public methods.

File: Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
+#ifndef SPIRE_STYLE_DROPDOWN_MENU_ITEM_HPP
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp]: Two words.

File: Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+#include "Spire/Ui/Ui.hpp"
+
+namespace Spire {
+
+  //! Represents a style item in a StyleDropDownMenu.
+  class StyleDropDownMenuItem : public QWidget {
+    public:
+
+      //! Signals that this item was selected.
+      using SelectedSignal = Signal<void (TrendLineStyle style)>;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp]: What is "this" item? Document the parameters to the slot.

File: Applications/Spire/Include/Spire/Charting/ColorPicker.hpp [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+#include "Spire/Charting/Charting.hpp"
+
+namespace Spire {
+
+  //! Displays a color picker and signals when a color is selected.
+  class ColorPicker : public QWidget {
+    public:
+
+      //! Signals that a color has been interacted with.
+      using ColorSignal = Signal<void (const QColor& color)>;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/ColorPicker.hpp]: Document the parameters to the signal. What does it mean to "interact" with a color?

File: Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+      //! Constructs a DropDownMenuItem with a specified style.
+      /*
+        \param style The style to display.
+        \param parent The parent to this widget.
+      */
+      StyleDropDownMenuItem(TrendLineStyle style, QWidget* parent = nullptr);
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp]: Mark as explicit.

**Comment by spireken** [comment]
Had to manually upload the build; it's called ChartingUiTester.

Have to update my upload script now.

File: Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+      //! Returns the style represented by this item.
+      TrendLineStyle get_style() const;
+
+      //! Sets the item's highlighted status.
+      void set_highlight();
+
+      //! Removes the item's highlighted status.
+      void remove_highlight();
+
+      boost::signals2::connection connect_selected_signal(
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp]: Document.

File: Applications/Spire/Include/Spire/Charting/StyleDropDownMenuList.hpp [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+#include "Spire/Ui/Ui.hpp"
+
+namespace Spire {
+
+  //! The list used by the StyleDropDownMenu.
+  class StyleDropDownMenuList : public QWidget {
+    public:
+
+      //! Signals that an item was selected from the list.
+      using SelectedSignal = Signal<void (TrendLineStyle style)>;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/StyleDropDownMenuList.hpp]: Documentation.

File: Applications/Spire/Include/Spire/Charting/StyleDropDownMenuList.hpp [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+      //! Constructs a StyleDropDownMenuList.
+      /*
+        \param parent The parent to the list.
+      */
+      StyleDropDownMenuList(QWidget* parent = nullptr);
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/StyleDropDownMenuList.hpp]: Mark explicit.

File: Applications/Spire/Include/Spire/Charting/StyleDropDownMenuList.hpp [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+      */
+      TrendLineStyle get_next(TrendLineStyle style);
+
+      //! Returns the item in the list that comes before the item specified.
+      /*
+        \param style The specified item.
+      */
+      TrendLineStyle get_previous(TrendLineStyle style);
+
+      boost::signals2::connection connect_selected_signal(
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/StyleDropDownMenuList.hpp]: Document.

File: Applications/Spire/Include/Spire/Charting/TrendLineEditor.hpp [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+      //! Signals that a line style has been selected.
+      using StyleSignal = Signal<void ()>;
+
+      //! Constructs a TrendLineEditor.
+      /*
+        \param parent The parent to the TrendLineEditor, and reference for
+                      determining the editor's position.
+      */
+      TrendLineEditor(QWidget* parent = nullptr);
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/TrendLineEditor.hpp]: Explicit.

File: Applications/Spire/Include/Spire/Charting/TrendLineEditor.hpp [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+      //! Returns the selected line style.
+      TrendLineStyle get_style() const;
+
+      //! Sets the selected line style.
+      /*
+        \param style The style to assign to the TrendLineEditor.
+      */
+      void set_style(TrendLineStyle style);
+
+      boost::signals2::connection connect_color_signal(
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/TrendLineEditor.hpp]: Document.

File: Applications/Spire/Include/Spire/Charting/TrendLineStyleDropDownMenu.hpp [hunk 15]
```diff
@@ -0,0 +1,40 @@
+#ifndef SPIRE_TREND_LINE_STYLE_DROPDOWN_MENU_HPP
+#define SPIRE_TREND_LINE_STYLE_DROPDOWN_MENU_HPP
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/TrendLineStyleDropDownMenu.hpp]: Two words.

File: Applications/Spire/Include/Spire/Charting/TrendLineStyleDropDownMenu.hpp [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+#include <QWidget>
+#include "Spire/Charting/Charting.hpp"
+#include "Spire/Charting/TrendLine.hpp"
+
+namespace Spire {
+
+  //! Displays drop down menu with selectable TrendLineStyles.
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/TrendLineStyleDropDownMenu.hpp]: Fix this sentence/typo.

File: Applications/Spire/Include/Spire/Charting/TrendLineStyleDropDownMenu.hpp [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+  class TrendLineStyleDropdownMenu : public QWidget {
+    public:
+
+      //! Constructs a TrendLineStyleDropDownMenu.
+      /*
+        \param parent The parent widget to the TrendLineStyleDropDownMenu.
+      */
+      TrendLineStyleDropdownMenu(QWidget* parent = nullptr);
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/TrendLineStyleDropDownMenu.hpp]: Explicit.

File: Applications/Spire/Include/Spire/Ui/DropDownMenu.hpp [hunk 18]
```diff
@@ -9,11 +9,11 @@ namespace Spire {
   class DropdownMenu : public QWidget {
     public:
 
-      //! Constructs a DropdownMenu with specified items. The first item
+      //! Constructs a DropDownMenu with specified items. The first item
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/DropDownMenu.hpp]: Fix grammar/typo.

File: Applications/Spire/Include/Spire/Ui/DropDownMenuItem.hpp [hunk 19]
```diff
@@ -5,19 +5,19 @@
...
 namespace Spire {
 
-  //! Represents an item in a DropdownMenu.
-  class DropdownMenuItem : public QLabel {
+  //! Represents an item in a DropDownMenu.
+  class DropDownMenuItem : public QLabel {
     public:
 
       //! Signals that this item was selected.
       using SelectedSignal = Signal<void (const QString& text)>;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Ui/DropDownMenuItem.hpp]: Document.

File: Applications/Spire/Source/Charting/ChartingWindow.cpp [hunk 20]
```diff
@@ -17,6 +17,7 @@
 #include "Spire/Ui/DropdownMenu.hpp"
 #include "Spire/Ui/SecurityWidget.hpp"
 #include "Spire/Ui/ToggleButton.hpp"
+#include "Spire/Charting/TrendLineEditor.hpp"
```
Comment (kranar) [inline Applications/Spire/Source/Charting/ChartingWindow.cpp]: Alphabetical order.

File: Applications/Spire/Source/Charting/DropDownColorPicker.cpp [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+
+using namespace boost::signals2;
+using namespace Spire;
+
+DropDownColorPicker::DropDownColorPicker(QWidget* parent)
+    : QWidget(parent),
+      m_has_mouse(false) {
+  setMouseTracking(true);
+  setFocusPolicy(Qt::StrongFocus);
+  window()->installEventFilter(this);
```
Comment (kranar) [inline Applications/Spire/Source/Charting/DropDownColorPicker.cpp]: Is there an alternative to installing an event filter.

File: Applications/Spire/Source/Charting/TrendLineStyleDropDownMenu.cpp [hunk 22]
```diff
@@ -0,0 +1,104 @@
+#include "Spire/Charting/TrendLineStyleDropDownMenu.hpp"
+#include <QLayout>
+#include <QPainter>
+#include <QPaintEvent>
+#include "Spire/Charting/StyleDropDownMenuList.hpp"
+#include "Spire/Spire/Dimensions.hpp"
+
+using namespace Spire;
+
+TrendLineStyleDropdownMenu::TrendLineStyleDropdownMenu(
```
Comment (kranar) [inline Applications/Spire/Source/Charting/TrendLineStyleDropDownMenu.cpp]: Rename to DropDown

File: Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp (hunk 6 above)
Comment (spireken) [reply Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp]: What would be better? That 'an instance of this class was selected'?

Comment (kranar) [reply Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp]: Signals a user selected a trend line.
param style - The trend line selected.

File: Applications/Spire/Source/Charting/DropDownColorPicker.cpp (hunk 21 above)
Comment (spireken) [reply Applications/Spire/Source/Charting/DropDownColorPicker.cpp]: It does't appear so. There's no signal to just listen for a resize or move.

Comment (kranar) [reply Applications/Spire/Source/Charting/DropDownColorPicker.cpp]: Okay leave as is.

File: Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp [hunk 23]
```diff
@@ -10,15 +10,19 @@ namespace Spire {
   class StyleDropDownMenuItem : public QWidget {
     public:
 
-      //! Signals that this item was selected.
+      //! Signals that an user selected a trend line style.
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp]: Fix typo on ```an```.

File: Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp [hunk 24]
```diff
@@ -10,15 +10,19 @@ namespace Spire {
...
+      /*
+        \param style The trend line selected.
+      */
       using SelectedSignal = Signal<void (TrendLineStyle style)>;
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/StyleDropDownMenuItem.hpp]: Is it a trend line that's selected or its style that's selected?

File: Applications/Spire/Include/Spire/Charting/StyleDropDownMenuList.hpp [hunk 25]
```diff
@@ -12,13 +12,16 @@ namespace Spire {
     public:
 
       //! Signals that an item was selected from the list.
+      /*
+        param style The trend line style that was selected.
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/StyleDropDownMenuList.hpp]: Fix this documentation.

File: Applications/Spire/Include/Spire/Charting/ColorPicker.hpp [hunk 26]
```diff
@@ -21,9 +24,17 @@ namespace Spire {
       */
       ColorPicker(int width, int height, QWidget* parent = nullptr);
 
+      //! Calls the provided slot when the color preview signal is triggered.
+      /*
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/ColorPicker.hpp]: Fix up all of these ```/*``` to ```/*!```

File: Applications/Spire/Include/Spire/Charting/DropDownColorPicker.hpp [hunk 27]
```diff
@@ -10,14 +10,21 @@ namespace Spire {
     public:
 
       //! Signals that a color has been selected.
+      /*
+        param color The color that was selected.
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Charting/DropDownColorPicker.hpp]: Fix the docstyle.

File: Applications/Spire/Include/Spire/Charting/ColorPicker.hpp (hunk 26 above)
Comment (spireken) [reply Applications/Spire/Include/Spire/Charting/ColorPicker.hpp]: Do you want me to fix all of them across the entire solution?

**Pull request merged**